File: .claude/skills/x-deep-scan/scripts/compute_stats.py
#!/usr/bin/env python3
"""Deterministic quantitative stats over a fetched tweet corpus.

Usage:
  python compute_stats.py --tweets-file raw/<handle>/<start>_<end>/tweets.json

Prints a JSON stats blob to stdout. Field names are read defensively since
different Apify actor versions rename fields -- unknown/missing values are
treated as 0/None rather than raising.
"""
import argparse
import datetime as dt
import json
import re
import statistics as st
import sys
from collections import Counter

URL_RE = re.compile(r"https?://\S+")


def g(item, *keys, default=None):
    for k in keys:
        if k in item and item[k] is not None:
            return item[k]
    return default


def parse_created_at(raw):
    if not raw:
        return None
    try:
        return dt.datetime.fromisoformat(raw.replace("Z", "+00:00"))
    except ValueError:
        pass
    try:
        return dt.datetime.strptime(raw, "%a %b %d %H:%M:%S %z %Y")
    except ValueError:
        return None


def has_media(item):
    for key in ("media", "extendedEntities", "entities"):
        val = item.get(key)
        if isinstance(val, dict) and val.get("media"):
            return True
        if isinstance(val, list) and val:
            return True
    return False


def has_link(item, text):
    ents = item.get("entities")
    if isinstance(ents, dict) and ents.get("urls"):
        return True
    return bool(URL_RE.search(text or ""))


def safe_mean_stdev(values):
    values = [v for v in values if v is not None]
    if not values:
        return {"mean": 0, "median": 0, "stdev": 0, "n": 0}
    mean = st.mean(values)
    median = st.median(values)
    stdev = st.stdev(values) if len(values) >= 2 else 0
    return {"mean": round(mean, 2), "median": round(median, 2), "stdev": round(stdev, 2), "n": len(values)}


def main():
    ap = argparse.ArgumentParser()
    ap.add_argument("--tweets-file", required=True)
    args = ap.parse_args()

    with open(args.tweets_file, "r", encoding="utf-8") as f:
        data = json.load(f)
    # Accept both a bare tweet array and the n8n fetcher's corpus wrapper {handle, range, tweets: [...]}
    tweets = data.get("tweets", []) if isinstance(data, dict) else data

    if not tweets:
        print(json.dumps({"tweet_count": 0, "warning": "no tweets in corpus"}))
        return

    rows = []
    for item in tweets:
        text = g(item, "text", "fullText", default="") or ""
        created = parse_created_at(g(item, "createdAt", "created_at"))
        rows.append({
            "id": g(item, "id", "tweetId"),
            "text": text,
            "created_at": created,
            "likes": g(item, "likeCount", "favoriteCount", default=0) or 0,
            "retweets": g(item, "retweetCount", default=0) or 0,
            "replies": g(item, "replyCount", default=0) or 0,
            "quotes": g(item, "quoteCount", default=0) or 0,
            "views": g(item, "viewCount", "views", default=None),
            "is_reply": bool(g(item, "isReply", default=False)),
            "is_retweet": bool(g(item, "isRetweet", default=False)),
            "is_quote": bool(g(item, "isQuote", default=False)),
            "has_media": has_media(item),
            "has_link": has_link(item, text),
            "char_len": len(text),
            "url": g(item, "url", "twitterUrl"),
        })

    dated = [r for r in rows if r["created_at"] is not None]
    if len(dated) < len(rows):
        undated_warning = f"{len(rows) - len(dated)} tweet(s) had unparseable createdAt and were excluded from cadence/time stats"
    else:
        undated_warning = None

    if dated:
        days_span = max(1, (max(r["created_at"] for r in dated) - min(r["created_at"] for r in dated)).days + 1)
        by_day = Counter(r["created_at"].date().isoformat() for r in dated)
        tweets_per_day = list(by_day.values())
        cadence = {
            "tweets_per_day_mean": round(len(dated) / days_span, 2),
            "tweets_per_day_stdev": round(st.stdev(tweets_per_day), 2) if len(tweets_per_day) >= 2 else 0,
            "days_span": days_span,
        }
        hour_hist = Counter(r["created_at"].hour for r in dated)
        weekday_hist = Counter(r["created_at"].strftime("%A") for r in dated)
    else:
        cadence = {"tweets_per_day_mean": 0, "tweets_per_day_stdev": 0, "days_span": 0}
        hour_hist, weekday_hist = Counter(), Counter()

    length_stats = safe_mean_stdev([r["char_len"] for r in rows])
    length_p90 = None
    if rows:
        sorted_lens = sorted(r["char_len"] for r in rows)
        idx = min(len(sorted_lens) - 1, int(0.9 * len(sorted_lens)))
        length_p90 = sorted_lens[idx]

    engagement = {
        "likes": safe_mean_stdev([r["likes"] for r in rows]),
        "retweets": safe_mean_stdev([r["retweets"] for r in rows]),
        "replies": safe_mean_stdev([r["replies"] for r in rows]),
        "quotes": safe_mean_stdev([r["quotes"] for r in rows]),
        "views": safe_mean_stdev([r["views"] for r in rows if r["views"] is not None]),
    }

    top10 = sorted(rows, key=lambda r: r["likes"], reverse=True)[:10]

    n = len(rows)
    format_flags = {
        "has_media": sum(1 for r in rows if r["has_media"]) / n,
        "has_link": sum(1 for r in rows if r["has_link"]) / n,
        "is_reply": sum(1 for r in rows if r["is_reply"]) / n,
        "is_retweet": sum(1 for r in rows if r["is_retweet"]) / n,
        "is_quote": sum(1 for r in rows if r["is_quote"]) / n,
    }
    format_flags = {k: round(v, 3) for k, v in format_flags.items()}

    out = {
        "tweet_count": n,
        "cadence": cadence,
        "length": {"char_mean": length_stats["mean"], "char_median": length_stats["median"],
                   "char_stdev": length_stats["stdev"], "char_p90": length_p90},
        "engagement": engagement,
        "format_flags": format_flags,
        "posting_hour_histogram": dict(hour_hist),
        "posting_weekday_histogram": dict(weekday_hist),
        "top10_by_likes": [
            {"id": r["id"], "url": r["url"], "text": r["text"][:280], "likes": r["likes"],
             "retweets": r["retweets"], "replies": r["replies"], "views": r["views"]}
            for r in top10
        ],
        "warnings": [w for w in [undated_warning] if w],
    }
    print(json.dumps(out, ensure_ascii=False, indent=2))


if __name__ == "__main__":
    main()

File: .claude/skills/x-deep-scan/scripts/compute_stats.js
#!/usr/bin/env node
// Deterministic quantitative stats over a fetched tweet corpus.
//
// Usage: node compute_stats.js --tweets-file <path>
//
// Accepts either a bare tweet array or the n8n fetcher's corpus wrapper
// {handle, range, tweets: [...]}. Field names are read defensively since
// Apify actor versions rename fields -- missing values become 0/null.

const fs = require('fs');

function arg(name) {
  const i = process.argv.indexOf(name);
  return i >= 0 ? process.argv[i + 1] : null;
}

function g(item, keys, dflt = null) {
  for (const k of keys) if (item[k] !== undefined && item[k] !== null) return item[k];
  return dflt;
}

function parseCreatedAt(raw) {
  if (!raw) return null;
  const d = new Date(raw);
  return isNaN(d.getTime()) ? null : d;
}

function hasMedia(item) {
  for (const key of ['media', 'extendedEntities', 'entities']) {
    const val = item[key];
    if (val && typeof val === 'object' && !Array.isArray(val) && val.media && val.media.length) return true;
    if (Array.isArray(val) && val.length) return true;
  }
  return false;
}

function hasLink(item, text) {
  const ents = item.entities;
  if (ents && typeof ents === 'object' && ents.urls && ents.urls.length) return true;
  return /https?:\/\/\S+/.test(text || '');
}

function stats(values) {
  const v = values.filter(x => x !== null && x !== undefined);
  if (!v.length) return { mean: 0, median: 0, stdev: 0, n: 0 };
  const mean = v.reduce((a, b) => a + b, 0) / v.length;
  const sorted = [...v].sort((a, b) => a - b);
  const mid = Math.floor(sorted.length / 2);
  const median = sorted.length % 2 ? sorted[mid] : (sorted[mid - 1] + sorted[mid]) / 2;
  const stdev = v.length >= 2 ? Math.sqrt(v.reduce((a, b) => a + (b - mean) ** 2, 0) / (v.length - 1)) : 0;
  const r = x => Math.round(x * 100) / 100;
  return { mean: r(mean), median: r(median), stdev: r(stdev), n: v.length };
}

function main() {
  const file = arg('--tweets-file');
  if (!file) { console.error('ERROR: --tweets-file required'); process.exit(1); }
  const data = JSON.parse(fs.readFileSync(file, 'utf-8'));
  const tweets = Array.isArray(data) ? data : (data.tweets || []);

  if (!tweets.length) { console.log(JSON.stringify({ tweet_count: 0, warning: 'no tweets in corpus' })); return; }

  const rows = tweets.map(item => {
    const text = g(item, ['text', 'fullText'], '') || '';
    return {
      id: g(item, ['id', 'tweetId']),
      text,
      created_at: parseCreatedAt(g(item, ['createdAt', 'created_at'])),
      likes: g(item, ['likeCount', 'favoriteCount'], 0) || 0,
      retweets: g(item, ['retweetCount'], 0) || 0,
      replies: g(item, ['replyCount'], 0) || 0,
      quotes: g(item, ['quoteCount'], 0) || 0,
      views: g(item, ['viewCount', 'views'], null),
      is_reply: Boolean(g(item, ['isReply'], false)),
      is_retweet: Boolean(g(item, ['isRetweet'], false)),
      is_quote: Boolean(g(item, ['isQuote'], false)),
      has_media: hasMedia(item),
      has_link: hasLink(item, text),
      char_len: text.length,
      url: g(item, ['url', 'twitterUrl']),
    };
  });

  const dated = rows.filter(r => r.created_at);
  const warnings = [];
  if (dated.length < rows.length) warnings.push(`${rows.length - dated.length} tweet(s) had unparseable createdAt and were excluded from cadence/time stats`);

  let cadence = { tweets_per_day_mean: 0, tweets_per_day_stdev: 0, days_span: 0 };
  const hourHist = {}, weekdayHist = {};
  if (dated.length) {
    const times = dated.map(r => r.created_at.getTime());
    const daysSpan = Math.max(1, Math.round((Math.max(...times) - Math.min(...times)) / 86400000) + 1);
    const byDay = {};
    for (const r of dated) {
      const day = r.created_at.toISOString().slice(0, 10);
      byDay[day] = (byDay[day] || 0) + 1;
      const h = r.created_at.getUTCHours();
      hourHist[h] = (hourHist[h] || 0) + 1;
      const wd = r.created_at.toLocaleDateString('en-US', { weekday: 'long', timeZone: 'UTC' });
      weekdayHist[wd] = (weekdayHist[wd] || 0) + 1;
    }
    const perDay = Object.values(byDay);
    const perDayStats = stats(perDay);
    cadence = {
      tweets_per_day_mean: Math.round((dated.length / daysSpan) * 100) / 100,
      tweets_per_day_stdev: perDay.length >= 2 ? perDayStats.stdev : 0,
      days_span: daysSpan,
    };
  }

  const lenStats = stats(rows.map(r => r.char_len));
  const sortedLens = rows.map(r => r.char_len).sort((a, b) => a - b);
  const p90 = sortedLens[Math.min(sortedLens.length - 1, Math.floor(0.9 * sortedLens.length))];

  const engagement = {
    likes: stats(rows.map(r => r.likes)),
    retweets: stats(rows.map(r => r.retweets)),
    replies: stats(rows.map(r => r.replies)),
    quotes: stats(rows.map(r => r.quotes)),
    views: stats(rows.filter(r => r.views !== null).map(r => r.views)),
  };

  const top10 = [...rows].sort((a, b) => b.likes - a.likes).slice(0, 10);
  const n = rows.length;
  const frac = f => Math.round((rows.filter(f).length / n) * 1000) / 1000;
  const formatFlags = {
    has_media: frac(r => r.has_media),
    has_link: frac(r => r.has_link),
    is_reply: frac(r => r.is_reply),
    is_retweet: frac(r => r.is_retweet),
    is_quote: frac(r => r.is_quote),
  };

  console.log(JSON.stringify({
    tweet_count: n,
    cadence,
    length: { char_mean: lenStats.mean, char_median: lenStats.median, char_stdev: lenStats.stdev, char_p90: p90 },
    engagement,
    format_flags: formatFlags,
    posting_hour_histogram: hourHist,
    posting_weekday_histogram: weekdayHist,
    top10_by_likes: top10.map(r => ({ id: r.id, url: r.url, text: r.text.slice(0, 280), likes: r.likes, retweets: r.retweets, replies: r.replies, views: r.views })),
    warnings,
  }, null, 2));
}

main();
